### ISSUE TITLE
Testing applications that use napari

### ISSUE THREAD

**Issue by sumanthratna** [napari/napari#2082]
## 📚 Documentation

I have a small package called [`viewmask`](https://github.com/sumanthratna/viewmask) that's a Click CLI that launches a napari viewer. I'm trying to write a really basic unit test to see if running `viewmask annotations nuclei.xml` will launch a napari viewer and have two layers. IMO there should be documentation on this, at least with a very basic example like I described. 

P.S. I don't actually know how to write the test that I described so please let me know, along with adding this to the docs (once I get comfortable with unit testing with napari, I can help write docs myself)

**Comment by tlambert03** [comment]
Hi @sumanthratna
You're absolutely right, we need to improve our docs for this!  It's definitely not a trivial thing to set up and we _would_ like to make it easy for you to test your code using napari.

The tips provided here are in no way meant to be "here you go... now you're on your own!".  But hopefully will give you a little something to go on until we can get around to writing more thorough docs for this.

We have a [pytest fixture](https://docs.pytest.org/en/stable/fixture.html) called `make_test_viewer` that you should be able to use for your own tests.  The basic idea is like this (but I haven't personally tested doing this outside of the napari repo itself):
```python
from napari.conftest import make_test_viewer

def test_viewer(self, make_test_viewer):
    """example of testing the viewer."""
    viewer = make_test_viewer()
    # do whatever you want with the viewer
 ```

[here](https://github.com/ome/ome-zarr-py/blob/8436d62c9ddf7009cff0cafb480b591941b74817/tests/test_napari.py) is an example of the  `ome-zarr-py` repo using our fixture for their tests.

note, this requires that you have `pytest-qt` installed, in addition to pytest.  If you're _totally_ new to writing unit tests in python, then it might help to look  through some of the [getting started](https://docs.pytest.org/en/stable/getting-started.html) documentation at pytest.

**Comment by tlambert03** [comment]
and also a small tip:  try to write more tests that cover smaller "units" of your code (many of which probably won't need to use the napari viewer at all), rather than going straight for the big full blown test that mimics how a user might interact with your program.

**Comment by jni** [comment]
@tlambert03 but I don't think the above will help @sumanthratna to test whether launching from CLI gives the expected result? But tbh at that point I don't know myself what to do... =\

**Comment by tlambert03** [comment]
if you want to test launching the full program from the cli, then you don't actually need anything from napari (like `make_test_viewer`).  I'd recommend looking through our [`test_cli.py`](https://github.com/napari/napari/blob/master/napari/_tests/test_cli.py) file, where we test our own cli interface.  There, the general testing tips to be aware about (and none of these are napari-specific):
- using [pytest monkeypatch fixture](https://docs.pytest.org/en/stable/monkeypatch.html) to provide `sys.argv` arguments
- using the [capsys fixture](https://docs.pytest.org/en/stable/capture.html) if you want to make assertions about stderr or stdout
- using [`unittest.mock.MagicMock`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.MagicMock) to mock objects (such as the napari.Viewer) that you aren't trying to directly test (but perhaps would like to make sure they are being called with certain arguments) ... (as we do [here](https://github.com/napari/napari/blob/master/napari/_tests/test_cli.py#L88))

**Comment by sofroniewn** [comment]
cc @neuromusic @ziyangczi @justinelarsen who are starting to think about "developer services" we could provide to help make this easier for people.

**Comment by sumanthratna** [comment]
Thanks for all the advice and sorry for the late reply! After looking at some of those links, I came across this function:
https://github.com/napari/napari/blob/da4e309e5734931a7ecfc821ae84a33462d80a45/napari/_qt/qt_event_loop.py#L38-L138

If I'm understanding the code correctly, I think this means that a global `Viewer` instance could be stored along with the global Qt app instance correct? I think that could make for a better DX, allowing something like this:
```python
    # ...
    def test_cli_napari(self):
        # running the CLI should launch a napari viewer with 3 layers

        # step 1: run the command; there are four options
        # - using Click's testing utils (may not apply if you're not using Click)
        # - use subprocess.run/Popen
        # - call the CLI function directly
        # - monkeypatch sys.argv

        # step 2: get napari's viewer instance; this method doesn't yet exist
        viewer = napari.get_viewer()

        # step 3: test what you want; `viewer` is just an instance of `napari.Viewer`
        assert len(viewer.layers) == 3
```
not sure if this is particularly dangerous, but it looks nice to me because it doesn't involve mocking a `napari.Viewer` which makes for a nice e2e test

**Comment by tlambert03** [comment]
> If I'm understanding the code correctly, I think this means that a global Viewer instance could be stored along with the global Qt app instance, correct?

It does mean that... but I don't think we _want_ to have a global singleton napari `Viewer` like that.  Right now, there's nothing preventing someone from creating multiple napari `Viewers` and that's a good thing.

however, for _your_ tests, if you know you're only ever going to have a single napari window, then you can search the QApp singeltons `topLevelWidgets()` like this:

```python
from qtpy.QtWidgets import QApplication, QMainWindow
from napari._qt.qt_viewer import QtViewer

# get all top level widgets
top_widgets = QApplication.instance().topLevelWidgets()
# find the MainWindow (this will raise StopIterationError if there isn't one)
main = next(widget for widget in top_widgets if isinstance(widget, QMainWindow))
# get the `napari.viewer.Viewer` instance from the main window
viewer = main.findChild(QtViewer).viewer
```

(I get that that's pretty ugly... and I do think we should probably save a reference to instantiated viewers somewhere, perhaps on our custom QMainWindow object?)

> it doesn't involve mocking a napari.Viewer which makes for a nice e2e test

just my opinion, but I think mocks are a good thing!  Since it isolates _your_ app's logic from your dependencies.  In this case, if you'd like to make sure your app added layers to the viewer, rather than checking the length of `viewer.layers`, good practice would be to check that your logic called (for example) `viewer.add_shapes` with a particular set of arguments, given certain command line arguments.  That's exactly what mocks are for, and the [`assert_called_with`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_with) function is your friend here:

```
# untested... but this is the general idea
viewer_mock = MagicMock(napari.Viewer)
monkeypatch.setattr(napari, 'Viewer', viewer_mock)
...
viewer.add_shapes.assert_called_with(...)
```

**Comment by sumanthratna** [comment]
> > it doesn't involve mocking a napari.Viewer which makes for a nice e2e test
> 
> just my opinion, but I think mocks are a good thing! Since it isolates _your_ app's logic from your dependencies. In this case, if you'd like to make sure your app added layers to the viewer, rather than checking the length of `viewer.layers`, good practice would be to check that your logic called (for example) `viewer.add_shapes` with a particular set of arguments, given certain command line arguments. That's exactly what mocks are for, and the [`assert_called_with`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_with) function is your friend here:
> 
> ```python
> # untested... but this is the general idea
> viewer_mock = MagicMock(napari.Viewer)
> monkeypatch.setattr(napari, 'Viewer', viewer_mock)
> ...
> viewer_mock.add_shapes.assert_called_with(...)
> ```
> 
> (linking to [this test](https://github.com/napari/napari/blob/da4e309e5734931a7ecfc821ae84a33462d80a45/napari/_tests/test_cli.py#L88-L101) again as an example)

You're right—that last example and everything else you've mentioned have changed my mind :) mocking `napari.Viewer` seems to be the way to go

I'm going to leave this issue open since it's a docs issue; if I ever get around to writing tests I'll come back with a docs PR that hopefully adheres to the practices you all have suggested. Thanks!

**Comment by Czaki** [comment]
@psobolewskiPhD @melissawm I think that it may be solved? If not, then should be moved to docs repo.

**Comment by psobolewskiPhD** [comment]
It's mostly? addressed by https://napari.org/stable/plugins/testing_workshop_docs/index.html
and https://napari.org/stable/plugins/test_deploy.html#tips-for-testing-napari-plugins
There's not much on `make_napari_viewer` and nothing on more advanced things like monkeypatching or using MagicMock that I can quickly find. At the same time, those arn't really napari specific...

Anyhow, I think some of the discussion above is already somewhat dated, and we have much of the relevant issues already in docs:
https://github.com/napari/docs/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+testing

So I'm going to transfer this over to docs and then we can close?
